Feature/gcg 38 admin page

Ag4Hz opened the pull request:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- Admin page implemented
- Only admins can access the Admin page
- On the Admin page admins can alter settings for level thresholds, skill multipliers and xp settings
- Settings are all saved to the database

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- role column added to users table, where the default value is 'user'

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added role column to the Vue User model

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added 'Admin' button to the navbar

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- created blank 'Admin' page
- created middleware for checking if the user has admin privileges
- created route to the 'Admin' page

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- created blank 'Admin' page
- created middleware for checking if the user has admin privileges
- created route to the 'Admin' page

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added count() func to count the total number of users to User.php

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- XPHelper.php extended to be more modular

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- navbar edited with Admin button which now has a ShieldCheck icon

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- ProfileController.php edited to make in the show function the user parameter to be nullable

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- created AdminController.php which makes queries to the db and returns the results

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- created Admin.vue to show a basic dashboard with all relevant user data from the db

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- Admin.vue refactored to use as many premade components as possible

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added multiplier column to skills table in db

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- multiplier column in skills table in db
- optimized AdminController

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added getXPConfigs to XPHelper.php

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- AdminController.php extended
- Admin.vue extended with XP settings, level thresholds and skill weights

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added level threshold edit mode

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added skill weight edit mode

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added xp settings edit mode

Ag4Hz committed:
format + lint fix

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- added migrations and models for GeneralSettings and LevelThresholds

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- refactored AdminController and XPHelper to use the implemented migrations and model logic

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- bugfixes to migrations

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- more bugfixes to controller logic and model structure

Ag4Hz committed:
GCG-38: As an admin, I want to configure XP rules so that rewards are tunable.

- migration bugfix

Ag4Hz committed:
GCG-38:

- corrected current level calculation and tweaked how the saving of settings works

Ag4Hz committed:
GCG-38:

- if cancel clicked, the changes revert

krisztina-tropee commented on app/Http/Controllers/AdminController.php:
```diff
@@ -0,0 +1,194 @@
+        ]);
+    }
+
+    private function getXPStats()
```
should be extracted into a separate file from controller

krisztina-tropee commented on app/Http/Controllers/AdminController.php:
```diff
@@ -0,0 +1,194 @@
+
+    public function updateAllSettings(Request $request)
+    {
+        $validated = $request->validate([
```
Please use form request class for the validation, the controller to remain clean

krisztina-tropee commented on app/Http/Controllers/AdminController.php:
```diff
@@ -0,0 +1,194 @@
+            // Update skill weights if provided
+            if (isset($validated['skill_weights'])) {
+                foreach ($validated['skill_weights'] as $skillData) {
+                    DB::table('skills')
```
n+1 query issue
if there can be many skill_weights, then worths refactoring to bulk update query

krisztina-tropee commented on app/Http/Controllers/AdminController.php:
```diff
@@ -0,0 +1,194 @@
+
+    public function updateThresholds(Request $request)
+    {
+        $thresholds = $request->validate([
```
extract it

krisztina-tropee commented on app/Http/Controllers/AdminController.php:
```diff
@@ -0,0 +1,194 @@
+
+    public function updateSkillWeights(Request $request)
+    {
+        $skillWeights = $request->validate([
```
extract it

krisztina-tropee commented on app/Http/Controllers/AdminController.php:
```diff
@@ -0,0 +1,194 @@
+            'skillWeights.*.multiplier' => 'required|numeric|min:0|max:10'
+        ]);
+
+        foreach ($skillWeights['skillWeights'] as $skillData) {
```
same n+1 quey issue if many skillWeights

krisztina-tropee commented on app/Http/Middleware/AdminMiddleware.php:
```diff
@@ -0,0 +1,23 @@
+     */
+    public function handle(Request $request, Closure $next): Response
+    {
+        if (auth()->user()->role !== 'admin') {
```
please use a php enum for the roles when you reference to them as a string

krisztina-tropee commented on app/Models/LevelThreshold.php:
```diff
@@ -0,0 +1,58 @@
+        return static::orderBy('level')->pluck('xp_required', 'level')->toArray();
+    }
+
+    public static function getThresholdForLevel(int $level): ?int
```
if function name is what you gave here then return the $threshold, otherwise better rename it to something which marks that you return the xp_required attribute

krisztina-tropee commented on resources/js/components/AppHeader.vue:
```diff
@@ -48,6 +48,13 @@ const mainNavItems: NavItem[] = [
     },
 ];
 
+if (auth.value.user.role === 'admin') {
```
define enum for role on jsvascript side as well

krisztina-tropee commented on resources/js/components/AppHeader.vue:
```diff
@@ -48,6 +48,13 @@ const mainNavItems: NavItem[] = [
 ];
 
+if (auth.value.user.role === 'admin') {
+    mainNavItems.push({
```
define mainNavItems as a computed and inside of it you can move this if

krisztina-tropee commented on routes/web.php:
```diff
@@ -35,10 +36,19 @@
+Route::middleware(['auth', 'verified', 'admin'])->group(function () {
+    Route::get('/admin', [AdminController::class, 'index'])->name('admin.index');
+});
+Route::post('/admin/thresholds', [AdminController::class, 'updateThresholds'])->name('admin.thresholds.update');
```
these why aren't under the admin middleware group too?

Ag4Hz committed:
GCG-38:

- fixed code review issues

Ag4Hz committed:
format

arnold-webgurus commented on app/Models/GeneralSetting.php:
```diff
@@ -0,0 +1,56 @@
+        'description',
+    ];
+
+    protected $casts = [
```
The created_at and updated_at columns are cast by default. But if you want to cast some attribute, use the `cast()` function instead.

arnold-webgurus commented on app/Models/GeneralSetting.php:
```diff
@@ -0,0 +1,56 @@
+    ];
+
+    // Accessor to get the properly typed value
+    protected function typedValue(): Attribute
```
Why do you need this casting?

arnold-webgurus commented on app/Models/LevelThreshold.php:
```diff
@@ -0,0 +1,52 @@
+        parent::boot();
+
+        static::creating(function ($threshold) {
+            if (static::where('level', $threshold->level)->exists()) {
```
Please use unique constraints on the database level.

Ag4Hz committed:
GCG-38:

- correction to how xp needed is calculated on the profile page

Ag4Hz committed:
GCG-38:

- minor fix to level distribution calculation

Ag4Hz committed:
GCG-38:

- unnecessary type column deleted from database

Ag4Hz committed:
GCG-38:

- another correction to the getLevelProgress function

Ag4Hz committed:
GCG-38:

- calculateLevel and getLevelProgress finally fixed

Ag4Hz committed:
GCG-38:

- removed casting of dates because they are cast by default
- removed typedValue

Ag4Hz committed:
GCG-38:

- removed casts entirely because it has no usage

krisztina-tropee commented on app/Models/LevelThreshold.php:
```diff
@@ -0,0 +1,41 @@
+        return static::orderBy('level')->pluck('xp_required', 'level')->toArray();
+    }
+
+    public static function updateThresholds(array $thresholds): void
```
would be great with bulk insert

Ag4Hz committed:
GCG-38:

- bulk insert instead of row by row in updateThresholds()

Ag4Hz committed:
Merge branch 'develop' into feature/GCG-38_AdminPage

krisztina-tropee commented on routes/web.php:
```diff
@@ -35,10 +36,19 @@
 });
 
+// Admin routes
+Route::middleware(['auth', 'verified', 'admin'])->group(function () {
```
if every route starts with 'admin.', then you could use a ->prefix on the Route as well

Ag4Hz committed:
GCG-38:

- naming conflict

Ag4Hz committed:
Merge branch 'feature/GCG-38_AdminPage' of https://github.com/Ag4Hz/GitCodeGud into feature/GCG-38_AdminPage

# Conflicts:
#	app/Http/Controllers/ProfileController.php
